pdns/auth-caches.cc

```diff
@@ -45,7 +47,31 @@ uint64_t purgeAuthCaches(const std::string& match)
   return ret;
 }
 
-/* remove specific entries from all caches, no wildcard matching */
+/* remove zone from all caches */
+uint64_t purgeAuthCacheForZone(const DNSName& name)
+{
+  uint64_t ret = 0;
+
+  if (g_zoneCache.isEnabled()) {
+    DomainInfo di;
+    UeberBackend B;
+    if (B.getDomainInfo(name, di, false)) {
+      // zone exists, add/update it in the zone cache.
+      // Handle this first, to avoid concurrent queries re-populating the other caches.
+      g_zoneCache.add(name, di.id);
```

Comment:
this takes a lock even if the zone cache already had the entry. Perhaps `g_zoneCache.add` could be smarter about that?